### ISSUE TITLE
Customise This! suggestion for lesson framework

### ISSUE THREAD

**Issue by ragamouf** [carpentries-incubator/docker-introduction#70]
Hello! This issue was raised at the recent [NZ Carpentries Connect unconference](https://pad.carpentries.org/ConnectUnconferenceNZ2021) See Line 159 - I’m tagging the following partners in crime who agreed we should submit this idea to the lesson: @dme26, @CallumWalley.  Also, thanks to @mpfl for drafting the background on this, because I am not an HPC specialist, but an enthusiastic enabler! 

Some infrastructure providers (e.g. HPC providers) do not run Docker due to security concerns in multi-tenanted environments, and prefer to run alternative container solutions like Singularity. The biggest problem is that Docker runs containers with root privileges by default.

Furthermore, some Linux flavours have moved away from using the Docker daemon by default and are using other container daemons like containerd.

Could there be a way to model a Customise This! component throughout the lesson to better support instructors customising this lesson for their own local environments? For example: including a checklist in the readme for instructors to be able to specify which container daemon to use, as well as local security requirements for the teaching environment.

**Comment by aturner-epcc** [comment]
@ragamouf Thanks for the ideas. I agree that it is particularly important to introduce Singularity when there is an HPC angle on the course. When @jcohen02 and I have run the Docker lesson in the past we actually run as a two day workshop with the [Singularity lesson](https://carpentries-incubator.github.io/singularity-introduction/) as there are usually a fair few learners that are HPC users. See, for example, https://epcced.github.io/2020-12-08-Containers-Online/ . This works well as we use a local Docker container to build the Singularity containers that we then deploy on the HPC system.

I also was also recently pointed at the excellent [Pawsey lesson on Docker and Singularity](https://pawseysc.github.io/singularity-containers/). I wonder if we should approach the developers of this lesson to see if we can bring the best bits of both sets of materials into the incubator lessons?

**Comment by ChristinaLK** [comment]
@ragamouf I think acknowledging the need for customization is important; I have mixed feelings about interspersing throughout the lesson. My main priority with this materials is to keep them novice-friendly, which means weeding out anything that doesn't contribute to the building of a mental model/basic skill practice. I'm also worried about cognitive overload. 
So my inclination would be to add something like a customize this section at the end, for people who want to apply these skills in a custom environment. 
That said, I can see the benefit to including notes about customizations throughout though, especially if learners *aren't* novices and will be asking these kind of questions all along.

**Comment by ragamouf** [comment]
@aturner-epcc That's interesting - I didn't know there was a Singularity lesson and it makes total sense you'd run them together. 

On the Pawsey lesson I can definitely reach out to @marcodelapierre to see if they'd be interested in collaborating or reviewing different lesson approaches. Maybe that's actually a good themed discussion, although I expect timezones may be uncomfortable! Pawsey folk are UTC+8; NZ folk are UTC+13 soon to be +12.

@ChristinaLK Thanks for this explanation - it's very useful to think about the likelihood of HPC learners being more polarised, and consider how to manage the room when novices and experts are in together. I think the original suggestion was looking at how to better support instructors first, in customising lessons to their local conditions. So the customisation flags would be added in the template, for instructors to adjust or take out as befits their audience. However, I defer to your appraisal because you know the lesson a lot more than I do. 

Tagging @dme26 @CallumWalley @mpfl and @Cholokiwi for your thoughts.

**Comment by jcohen02** [comment]
Just adding my thoughts to this discussion. Thanks for raising this issue @ragamouf. 

I can certainly see that there is scope for offering instructors some form of guidance on where or how to make customisations to the lesson in order to support different local environments or container platforms. At the same time, I suppose it depends on the amount of customisation that is required as to whether this ends up being something that turns into a substantial rewrite of the lesson content, or whether it adds confusion for the course participants.

@aturner-epcc and I have now had a couple of experiences of running the Singularity lesson where we've run on different underlying HPC platforms. In that case, we're still using a standard Singularity installation but the lesson needs to be customised to provide details on accessing the HPC platform and also on any configuration differences with Singularity on the target platform. This tends to require a bit of work but it's mostly confined to the start of the lesson and the setup instructions.

Nonetheless, I do agree with @ChristinaLK's point about keeping things novice friendly and avoiding information overload. I guess that the core content of the lesson is likely to be the same regardless of the underlying platform being covered, especially with libraries such as `containerd` offering a common interface/API to core container funcitonality. On this basis, adding some guidance for instructors looking to use the lesson material in the context of different target platforms or environments, perhaps in a customisation section, or notes for instructors, seems like a good idea.

Finally, I wasn't previously aware of the [Pawsey Docker and Singularity lesson](https://pawseysc.github.io/singularity-containers/) that @aturner-epcc highlighted. I think it would be a good idea to approach the team behind that lesson and see if there's scope to combine our efforts to enhance the materials in the incubator.

**Comment by ChristinaLK** [comment]
One final note: I think of this lesson specifically NOT as an HPC lesson b/c while many people do use containers on HPC/cloud systems, there are use cases outside of those systems and this whole lesson can be done on your laptop, no external system needed. 

In my mind, containers on HPC would be a "part 2" of this lesson; but what we have now is what's needed for people to get familiar with Docker containers on their own computer. So maybe some optional final episodes would be the way to go.  Or maybe we explicitly make this lesson about "containers on shared resources" - something that the maintainers can discuss when we meet in...April? Can't remember the exact date.

**Comment by sstevens2** [comment]
I see a reference to it in the maintainers meeting issues #50 as well.  I'll be sure the issues raised here get into the agenda for the March 16th meeting.

**Comment by marcodelapierre** [comment]
Hi everyone, thanks for mentioning me in this very interesting conversation :-) 

First thing: we are certainly keen to share our materials on Docker&Singularity with this Carpentry project! (with acknowledgement to the Pawsey Centre)
Right now I am unsure on how much of my time I can allocate in direct work, but I am certainly keen to contribute to discussion and share any useful experience. I am positive that I can make it, time zone-wise, to arrange a meeting with NZ and East Oz.

In terms of customisation, we started about 3 years ago with running Docker workshops, with a focus on Clodu, see for instance:
https://pawseysc.github.io/docker-resbaz19/
https://pawseysc.github.io/container-workflows/
And then I later developed the Singularity materials that are cited above.
Although there are other container runtimes out there, I think that right now Docker and Singularity together cover quite a large fraction of the overall container world in scientific computing. 
I tried to cover both in the same materials, and it did get kind of confusing. Some topics are the same, but others are different: 
* Docker comes from the IT world and one needs to spend lots of time in explaining the often cumbersome syntax required for running; on the other hand, there are no examples for HPC due to zero uptake in the space. Notably, as mentioned above, the image build part with Docker is quite useful for Singularity users, too, as Docker images are universal and thus more popular in the community, as opposed to Singularity images.
* Singularity as a simpler and more intuitive syntax, and you can then focus on HPC examples (MPI codes, GPU, ..). Note that Singularity can also be used on the cloud, so one could also try and suggest a model where Docker is only used for building on laptop or cloud, and then the image gets deployed and run everywhere with Singularity.

Long story short, probably a good approach is to maintain two distinct sets of materials for Docker and Singularity.

In terms of site specific customisation of the materials due to login/access and so on, we typically end up having an unrelated welcome page for the workshop, which details these aspects, and keep the bulk of the lesson materials unchanged.
In my experience, there were only a limited number of actual runtime examples that required site customisation: accessing Jupyter/RStudio web pages deployed as containers (firewall rules for access), and GPU codes (you typically need to go through the system job manager, i.e. Slurm, PBS or similar).

These are just some first thoughts I had, happy to follow-up with a more detailed conversation.

**Comment by ChristinaLK** [comment]
thanks @marcodelapierre !!

**Comment by CallumWalley** [comment]
Bit late to the party here, but thought I'd chip in regarding the customisation/localisation.

I liked the small amount of localisation in the [HPC carpentry ](https://github.com/CallumWalley/hpc-shell/tree/gh-pages/_includes/snippets/01) material, could do something similar for containers.

The default `_config.yml`  values can then be overwritten when serving, e.g.

```
--config \
    _config.yml,\
    _includes/_config_SLURM.yml,\
    _includes/_config_Singularity.yml,\
    _includes/_config_NeSI.yml,\
    _includes/_config_NeSI_Mahuika.yml
```

Would love to see more use of variables, tags, filters so this kind of thing can be more viable. 

p.s. I'm pretty new to the lesson template, so please let me know if there is a better way. :)

**Comment by aturner-epcc** [comment]
@CallumWalley Thanks for sharing. Personally, I do not think trying to make this lesson customisable with respect to container technology is the right way to go. One of the key things about this lesson is that people go away with a working Docker framework on their own system that they can continue to use to explore and work with in their area of interest. I am not sure what customisation would be useful in this lesson itself as it is focussed on using Docker on your own local system? (Maybe you could provide examples tailored to different research communities?)

There is more scope/requirement for customisation in the [Singularity lesson](https://carpentries-incubator.github.io/singularity-introduction/) as, typically, you are using Singularity on a particular remote computing platform and there are differences in scheduler setup, MPI setup etc. so I think some of the stuff from HPC Carpentry would work well there. (In fact, I think there is a reasonably strong argument that the Singularity lesson should be in HPC Carpentry but that is a different issue.)

**Comment by sstevens2** [comment]
Agree with @aturner-epcc.   I think it is good that this works on local machines and then the learners/instructors can customize for other environments as needed.

I think there are two options were would add additional customizable options.
1. High tech - Wait until we transition to the new lesson template and then look into how to do this there.
2. Low tech - Add info on customizing (or linking to custom forks) in the instructor notes

**Comment by ChristinaLK** [comment]
David to review and either close or create action item to close.

**Comment by dme26** [comment]
Thanks all for the discussion and contributions on this thread. My feelings, hopefully also appropriate recollecting and representing discussions at the maintainers' meetings, are:
* The core episodes for this lesson should focus on on-learners' laptops use, and Docker, however...
* The refactoring of material between episodes and extras allows for episodes that head off toward HPC, cloud and other target domains (also relevant is the development of "learner pathways" for the lesson, currently under development).
* Using "extra" episodes might avoid too much reliance on technology from the Carpentries template (sstevens2's "Low tech" option above) not least noting...
* The new Carpentries template will likely offer different possibilities for customisation, so we can review customisation when after the lesson has migrated to this template. (sstevens2's "High tech" option above)
* (One aspect that's interested me is whether we can at least highlight to workshop organisers what's likely to need to change, but I think I'll park that thought behind waiting for the new Carpentries template to see what mechanisms might be available to help.)
* The incubating Singularity lesson is now mentioned in `index.md` and `README.md` and briefly in episode 8.

So I *think* that given all that I'll close this issue for now, but fell free to reopen it, of course.
